src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -331,47 +331,56 @@ internal static List<CompletionResult> MakeCommandsUnique(IEnumerable<PSObject>
-                            if (!ModuleCmdletBase.IsPrefixedCommand(commandInfo))
-                            {
-                                completionName = commandInfo.ModuleName + "\\" + completionName;
-                            }
+                            continue;
                         }
-                    }
 
-                    results.Add(GetCommandNameCompletionResult(completionName, commandList[0], addAmpersandIfNecessary, quote));
+                        _ = modulesWithCommand.Add(commandInfo.ModuleName);
+                        if ((commandInfo.CommandType != CommandTypes.Alias && commandInfo.CommandMetadata.CommandType is not null)
+                            || (commandInfo.CommandType == CommandTypes.Alias && commandInfo.Definition is not null))
+                        {
+                            _ = importedModules.Add(commandInfo.ModuleName);
+                        }
+                    }
 
-                    // For the other commands that are hidden, we need to disambiguate,
-                    // but put these at the end as it's less likely any of the hidden
-                    // commands are desired.  If we can't add anything to disambiguate,
-                    // then we'll skip adding a completion result.
-                    for (int index = 1; index < commandList.Count; index++)
+                    int moduleCount = modulesWithCommand.Count;
+                    modulesWithCommand.Clear();
+                    int index;
+                    if (commandInfoArray[0].CommandType == CommandTypes.Application
+                        || importedModules.Count == 1
+                        || moduleCount < 2)
                     {
-                        var commandInfo = commandList[index] as CommandInfo;
-                        Diagnostics.Assert(commandInfo != null, "Elements should always be CommandInfo");
+                        // We can use the short name for this command because there's no ambiguity about which command it resolves to.
+                        index = 1;
+                        results.Add(GetCommandNameCompletionResult(keyValuePair.Key, commandInfoArray[0], addAmpersandIfNecessary, quote));
+                        _ = modulesWithCommand.Add(commandInfoArray[0].ModuleName);
+                    }
+                    else
+                    {
+                        index = 0;
+                    }
 
+                    for (; index < commandInfoArray.Length; index++)
+                    {
+                        CommandInfo commandInfo = commandInfoArray[index];
                         if (commandInfo.CommandType == CommandTypes.Application)
                         {
-                            endResults.Add(GetCommandNameCompletionResult(commandInfo.Definition, commandInfo, addAmpersandIfNecessary, quote));
+                            results.Add(GetCommandNameCompletionResult(commandInfo.Definition, commandInfo, addAmpersandIfNecessary, quote));
```

Comment:
Hmm, you are not using `endResults` anymore, meaning that you are not putting all the module-name-prefixed results at the end. Can you confirm if that expected and why? Also, if it's expected, you should remove the `endResults` declaration above and the use of it below.

Reply:
Yes that's intentional. It seemed weird to move commands with the same name far apart from each other so the user has to tab through the entire list to find the command alternatives. For example if I want the VMWare version of `Get-VMHost` and I type in `Get-VMH<Tab>` and end up with the Hyper-V version, why shouldn't the next result then be the VMWare version?